lib/fs/datablock_manager.cpp

```diff
@@ -0,0 +1,136 @@
+#include "fs.hpp"
+
+DataBlock_Manager::DataBlock_Manager(Fs *fs, u_int64_t block_segment_start,
+                                         u_int64_t block_segment_end)
+    : fs(fs), block_segment_start(block_segment_start),
+      block_segment_end(block_segment_end) {}
+
+class BitmapBlock_Data {
+public:
+  char buf[IO_BLOCK_SIZE];
+  u_int64_t datablocks_per_bitmap;
+
+  BitmapBlock_Data(u_int64_t datablocks_per_bitmap_)
+      : datablocks_per_bitmap(datablocks_per_bitmap_) {}
+
+  u_int64_t get_next_node() {
+    u_int64_t block_num;
+    read_u64(&block_num, buf);
+    return block_num;
+  }
+  void set_next_node(u_int64_t block_num) { write_u64(block_num, buf); }
+
+  u_int64_t find_unfilled() {
+    const char *data = &buf[8];
+    u_int64_t i = 0;
+
+    for (; i < datablocks_per_bitmap; ++i)
+      if ((data[i / 8] & (1 << (i % 8))) == 0)
+        return i + 1;
+
+    return 0;
+  }
+  u_int64_t claim_relative_block() {
+    u_int64_t unfilled = find_unfilled();
+    if (unfilled)
+      buf[((unfilled - 1) / 8) + 8] |= (1 << ((unfilled - 1) % 8));
+    return unfilled;
+  }
+
+  void release_relative_block(u_int64_t relative_block_num) {
+    relative_block_num -= 1;
+    size_t index = (relative_block_num / 8) + 8;
+    int offset = relative_block_num % 8;
+    buf[index] &= ~(1 << offset);
+  }
+};
+
+int DataBlock_Manager_Bitmap::new_datablock(u_int64_t *block_num) {
+  int err;
+  BitmapBlock_Data bitmap = BitmapBlock_Data(DATABLOCKS_PER_BITMAP_BLOCK);
+  u_int64_t bitmap_block_num = fs->superblock.free_list_head;
+  char zero_buf[IO_BLOCK_SIZE] = {0};
+
+  if (bitmap_block_num < block_segment_start ||
+      bitmap_block_num >= block_segment_end)
+    return -1;
+
+  if ((err = fs->disk->read_block(bitmap_block_num, bitmap.buf)) < 0)
+    return err;
+
+  u_int64_t relative_block_num = bitmap.claim_relative_block();
+
+  if (relative_block_num == 0)
+    return -1;
+
+  u_int64_t block_num_ = relative_block_num + bitmap_block_num;
+
+  // NOTE: this could be removed for speed
+  if ((err = fs->disk->write_block(block_num_, zero_buf)) < 0)
+    return err;
+
+  // Could be optimized
+  if (bitmap.find_unfilled() == 0) {
+    if ((err = fs->save_free_list_head(bitmap.get_next_node())) < 0)
+      return err;
+    bitmap.set_next_node(0);
+  }
+
+  if ((err = fs->disk->write_block(bitmap_block_num, bitmap.buf)) < 0)
+    return err;
+
+  (*block_num) = block_num_;
+  return 0;

```

Comment:
this changes the return behavior. should we make it return the block num?

Reply:
I am thinking that we may want to change the error handling implementation to work in a similar way to errno, in which case there would be no issue with returning the block number and 0 signify an error has occurred.